services/image-rec-master/service/app.py

```diff
@@ -0,0 +1,213 @@
+    get_list('unremoved-targets')
+    get_list('removing-targets')
+    get_set('removed-targets')
+
+    data_str = await tr.execute()
+
+    msg = PipelineState()
+    msg.time = time()
+
+    fields = [
+        msg.all_images,
+        msg.unprocessed_auto,
+        msg.processing_auto,
+        msg.processed_auto,
+        msg.retrying_auto,
+        msg.errored_auto,
+        msg.skipped_auto,
+        msg.unprocessed_manual,
+        msg.processed_manual,
+        msg.skipped_manual,
+        msg.all_targets,
+        msg.unsubmitted_targets,
+        msg.submitting_targets,
+        msg.submitted_targets,
+        msg.errored_targets,
+        msg.unremoved_targets,
+        msg.removing_targets,
+        msg.removed_targets
+    ]
+
+    # Convert the string lists to integer lists and add them to the
+    # proto message.
+    for field, str_list in zip(fields, data_str):
+        field.extend([int(id_str) for id_str in str_list])
+        field.sort()
+
+    return _proto_response(request, msg)
+
+
+@routes.get(r'/api/pipeline/targets/{id:\d+}')
+async def handle_get_pipeline_target_by_id(request):
+    """Return a target by id in the pipeline."""
+    target_id = int(request.match_info['id'])
+    target = await _get_target(request, target_id)
+
+    if target:
+        return _proto_response(request, target)
+    else:
+        return web.Response(status=404)
+
+
+@routes.post('/api/pipeline/reset')
+async def handle_reset_pipeline(request):
+    """Empty out the current Redis database to reset the pipeline."""
+    await request.app['redis'].flushdb()
+    return web.Response(status=204)
+
+
+@routes.get('/api/pipeline/archive')
+async def handle_get_pipeline_archive(request):
+    """Return an archive for backup target submission."""
+    odlcs = []
+    target_ids = await get_int_set(request.app['redis'], 'all-targets')
+
+    for target_id in target_ids:
+        target = await _get_target(request, target_id)
+
+        # Add the target if it hasn't errored or been removed.
+        if not target.errored and not target.removed:
+            odlcs.append(target.odlc)
+
+    archive = create_archive(odlcs)
+
+    if archive:
+        return web.Response(body=archive, headers={
+            'Content-Disposition': 'attachment; filename="targets.zip"',
+            'Content-Type': 'application/zip'
+        })
+    else:
+        return web.Response(status=204)
+
+
+async def _get_target(request, target_id):
+    target_hash = await request.app['redis'].hgetall(f'target:{target_id}')
+
+    if target_hash:
+        msg = Target()
+        msg.time = time()
+
+        msg.id = int(target_hash.get(b'id', b'0'))
+        msg.image_id = int(target_hash.get(b'image_id', b'0'))
+        msg.odlc.ParseFromString(target_hash.get(b'odlc', b''))
+        msg.submitted = target_hash.get(b'submitted', b'0') == b'1'
+        msg.errored = target_hash.get(b'errored', b'0') == b'1'
+        msg.removed = target_hash.get(b'removed', b'0') == b'1'
```

Comment:
Strings have to be bytes objects when working with aioredis? yikers, i've never had that problem

Reply:
Most Redis clients in languages treat everything going into Redis as binary data I believe?